hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -3242,13 +3244,16 @@ private long getTimeLimit(HBaseRpcController controller, boolean allowHeartbeatM
       if (controller != null && controller.getCallTimeout() > 0) {
         timeLimitDelta = Math.min(timeLimitDelta, controller.getCallTimeout());
       }
+      long now = EnvironmentEdgeManager.currentTime();
+      // if we can, subtract time already spent in rpc queue. otherwise we may still timeout
+      if (rpcCall != null) {
+        timeLimitDelta -= (now - rpcCall.getReceiveTime());
```

Comment:
Later we will use timeLimitDelta / 2, so I do not think we should subtract the queue time here. It does not make sense to count the remaining time and then divided by 2...

Reply:
Thanks for taking a look!

Why does that not make sense? The goal of the time limit is to return before the timeout. Dividing by 2 is not enough here. One could have a timeout of 5s and then be in the queue of 4s. In that case 5 / 2 = 2.5, but there's only 1s left. 

Just subtracting the queue time alone is not enough either because it would ride way too close to the timeout and could still timeout. If we returned 1s here but due to clock drift and it things there was only 900ms, we'd still time out. 

So I think subtracting queue time and then dividing by 2 makes sense. But if you think that is too much we could divide by a smaller factor instead. We really just need some buffer, in addition to counting queue time.

Reply:
I added this PR because we've specifically seen cases where the scanner still timeouts, even with dividing by 2. This only happens when server is overloaded with high queue times, but it exacerbated load issues and we should try to respond even under heavy load. 

It's especially bad if it happens on openScanner because the client will retry and the scanner will remain open. We've seen thousands of leaked scanners from this and server slows to crawl. I plan to push a separate PR later to try to close timed out scanners. But I think first we should try to account for queue times in time limit so we don't timeout at all if possible.

Reply:
I do not mean we should not count queue time, I just mean, we should not subtract it here. The queue is related to rpcTimeout, not related to scannerLeaseTimeoutPeriod right? So for me I prefer we substract it earlier, when calculate the rpcTimeout or callTimeout.

Reply:
Oh I see. I will try moving it up a few lines so it just applies to callTimeout or rpcTimeout. Thanks for clarifying.